File: src/sdks/typescript/src/dataplane.ts
import type { components } from './api/schema.gen'
import { AuthenticationError, FissionPlaneError } from './errors'
import type { Logger, RequestDefaults } from './http'

/** The well-known data-plane port the sandbox agent listens on. */
export const DEFAULT_AGENT_PORT = 50000

type CapabilityToken = components['schemas']['CapabilityToken']

/** The sandbox identity and current capability token used by data-plane modules. */
export interface TokenSource {
  readonly info: { readonly sandbox_id: string; readonly domain: string }
  readonly token?: CapabilityToken
  /**
   * Mints a token for the current epoch and arms the handle with it.
   * When present, data-plane modules recover from a rejected token by
   * calling it once and replaying the request.
   */
  readonly mintToken?: () => Promise<CapabilityToken>
}

/** Minimal WebSocket surface accepted by the SDK's streaming modules. */
export interface WebSocketConnection {
  readonly protocol: string
  onOpen?: () => void
  onMessage?: (data: unknown) => void
  onClose?: () => void
  onError?: (error: unknown) => void
  send(data: string): void
  close(code?: number, reason?: string): void
}

/** Injectable WebSocket factory used by process attach and file watch. */
export type WebSocketFactory = (
  url: string,
  protocols: string[],
) => WebSocketConnection

/** How sandbox handles reach the sandbox data plane. */
export interface DataPlaneOptions extends RequestDefaults {
  /**
   * Port the sandbox agent serves the data plane on; the first label of
   * the data-plane hostname.
   *
   * @default 50000
   */
  agentPort?: number
  /** Custom fetch implementation (tests, non-Node runtimes). */
  fetch?: typeof globalThis.fetch
  /** Custom WebSocket factory (tests, runtimes without a global WebSocket). */
  webSocket?: WebSocketFactory
}

class NativeWebSocketConnection implements WebSocketConnection {
  onOpen?: () => void
  onMessage?: (data: unknown) => void
  onClose?: () => void
  onError?: (error: unknown) => void

  constructor(private readonly socket: WebSocket) {
    socket.addEventListener('open', () => this.onOpen?.())
    socket.addEventListener('message', (event) => this.onMessage?.(event.data))
    socket.addEventListener('close', () => this.onClose?.())
    socket.addEventListener('error', (event) => this.onError?.(event))
  }

  get protocol(): string {
    return this.socket.protocol
  }

  send(data: string): void {
    this.socket.send(data)
  }

  close(code?: number, reason?: string): void {
    this.socket.close(code, reason)
  }
}

function defaultWebSocketFactory(
  url: string,
  protocols: string[],
): WebSocketConnection {
  if (typeof globalThis.WebSocket === 'undefined') {
    throw new FissionPlaneError(
      'WebSocket is unavailable; provide DataPlaneOptions.webSocket',
    )
  }
  return new NativeWebSocketConnection(new globalThis.WebSocket(url, protocols))
}

/**
 * Runs a data-plane call, recovering once from a rejected capability
 * token by minting a fresh one and replaying the call.
 *
 * A token is scoped to a sandbox epoch and expires, so a handle that
 * has been idle can hold one the agent no longer accepts. Every
 * data-plane HTTP call in `commands` and `files` goes through here.
 *
 * WebSocket streams are not covered: the handshake's 401 reaches the
 * SDK as an opaque socket error, so `commands.attach` and `files.watch`
 * still need a token minted in advance.
 *
 * @param sandbox Source of the sandbox identity and current token.
 * @param logger Diagnostic sink for the re-mint.
 * @param run Performs the call against the current token.
 * @returns The call's result, from the first or the replayed attempt.
 */
export async function withFreshToken<T>(
  sandbox: TokenSource,
  logger: Logger,
  run: () => Promise<T>,
): Promise<T> {
  try {
    return await run()
  } catch (error) {
    if (
      !(error instanceof AuthenticationError) ||
      sandbox.mintToken === undefined
    ) {
      throw error
    }
    logger.debug?.(
      `sandbox ${sandbox.info.sandbox_id} rejected the capability token; ` +
        'minting a fresh one and replaying the request',
    )
    await sandbox.mintToken()
    return run()
  }
}

/** Resolves the current token or raises the standard unarmed-handle error. */
export function capabilityToken(sandbox: TokenSource): string {
  const token = sandbox.token
  if (token === undefined) {
    throw new FissionPlaneError(
      `sandbox ${sandbox.info.sandbox_id} has no capability token; ` +
        'obtain one via sandboxes.create, sandbox.resume, or sandbox.mintToken',
    )
  }
  return token.token
}

/** Returns the HTTPS origin for one sandbox's data plane. */
export function dataPlaneOrigin(
  sandbox: TokenSource,
  agentPort: number,
): string {
  return `https://${agentPort}-${sandbox.info.sandbox_id}.${sandbox.info.domain}`
}

function base64Url(value: string): string {
  const bytes = new TextEncoder().encode(value)
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)
  return btoa(binary)
    .replaceAll('+', '-')
    .replaceAll('/', '_')
    .replace(/=+$/u, '')
}

/** Opens an authenticated data-plane WebSocket using protocol credentials. */
export function openDataPlaneWebSocket(
  sandbox: TokenSource,
  agentPort: number,
  factory: WebSocketFactory | undefined,
  path: string,
  query: Record<string, string | number | boolean | undefined>,
): WebSocketConnection {
  const token = capabilityToken(sandbox)
  const url = new URL(path, dataPlaneOrigin(sandbox, agentPort))
  url.protocol = 'wss:'
  for (const [name, value] of Object.entries(query)) {
    if (value !== undefined) url.searchParams.set(name, String(value))
  }
  return (factory ?? defaultWebSocketFactory)(url.toString(), [
    'fissionplane.v1',
    `fissionplane.token.${base64Url(token)}`,
  ])
}

File: src/sdks/rust/src/dataplane.rs
//! Shared transport for a sandbox's data plane.
//!
//! Every data-plane call carries a capability token bound to the
//! sandbox's epoch, and both can move underneath a long-lived handle:
//! the token expires, or a resume advances the epoch. Rather than
//! failing closed and making the caller rebuild the handle, each request
//! here re-mints once through the control plane when the agent answers
//! 401 and then replays the call with the fresh token.

use reqwest::header::{HeaderMap, HeaderValue};
use reqwest::{Method, Response};
use serde::de::DeserializeOwned;
use tokio_tungstenite::tungstenite;
use url::Url;

use crate::client::AgentTarget;
use crate::error::Error;
use crate::http::Http;
use crate::models;
use crate::streaming::{self, EventSocket};
use crate::token::TokenSource;

/// The header carrying a capability token on the data plane.
const TOKEN_HEADER: &str = "X-Sandbox-Token";

/// The agent's base URL for one sandbox, plus its refreshable
/// credential.
#[derive(Clone, Debug)]
pub(crate) struct DataPlane {
    http: Http,
    base: Url,
    credential: TokenSource,
}

impl DataPlane {
    pub(crate) fn new(
        http: &Http,
        agent: &AgentTarget,
        info: &models::Sandbox,
        credential: TokenSource,
    ) -> Result<Self, Error> {
        let base = agent_base_url(agent, info)?;
        Ok(Self {
            // The token is attached per request instead of baked into
            // the default headers, so a refresh takes effect at once.
            http: http.with_target(base.clone(), HeaderMap::new()),
            base,
            credential,
        })
    }

    /// The agent's base URL, for building stream URLs.
    pub(crate) fn base(&self) -> &Url {
        &self.base
    }

    pub(crate) fn request(&self, method: Method, path: &str) -> reqwest::RequestBuilder {
        self.http.request(method, path)
    }

    pub(crate) async fn send_json<T: DeserializeOwned>(
        &self,
        request: reqwest::RequestBuilder,
    ) -> Result<T, Error> {
        let response = self.execute(request).await?;
        Ok(response.json().await?)
    }

    pub(crate) async fn send_no_content(
        &self,
        request: reqwest::RequestBuilder,
    ) -> Result<(), Error> {
        self.execute(request).await?;
        Ok(())
    }

    pub(crate) async fn send_bytes(
        &self,
        request: reqwest::RequestBuilder,
    ) -> Result<Vec<u8>, Error> {
        let response = self.execute(request).await?;
        Ok(response.bytes().await?.to_vec())
    }

    /// Sends the request with the current token, re-minting once and
    /// replaying it if the agent rejects the token.
    async fn execute(&self, request: reqwest::RequestBuilder) -> Result<Response, Error> {
        let replay = request.try_clone();
        let token = self.credential.current();
        let first = self.http.execute(authorized(request, &token.token)?).await;
        match (first, replay) {
            (Err(rejected), Some(replay)) if is_unauthorized(&rejected) => {
                let token = self.credential.refresh().await?;
                self.http.execute(authorized(replay, &token.token)?).await
            },
            (outcome, _) => outcome,
        }
    }

    /// Opens a data-plane stream, re-minting once and reconnecting if
    /// the handshake is rejected for the token.
    pub(crate) async fn connect(
        &self,
        url: Url,
        query: &[(&str, String)],
    ) -> Result<EventSocket, Error> {
        let timeout = self.http.limits().request_timeout;
        let token = self.credential.current();
        match streaming::connect(url.clone(), &token.token, query, timeout).await {
            Err(rejected) if is_unauthorized_handshake(&rejected) => {
                let token = self.credential.refresh().await?;
                streaming::connect(url, &token.token, query, timeout).await
            },
            outcome => outcome,
        }
    }
}

/// The agent's base URL: the override when configured, otherwise the
/// agent port's own hostname.
pub(crate) fn agent_base_url(agent: &AgentTarget, info: &models::Sandbox) -> Result<Url, Error> {
    if let Some(base) = &agent.base_override {
        return Ok(base.clone());
    }
    let url = format!("https://{}-{}.{}", agent.port, info.sandbox_id, info.domain);
    Url::parse(&url).map_err(|source| Error::Config(format!("data-plane URL {url:?}: {source}")))
}

fn authorized(
    request: reqwest::RequestBuilder,
    token: &str,
) -> Result<reqwest::RequestBuilder, Error> {
    let mut value = HeaderValue::from_str(token)
        .map_err(|source| Error::Config(format!("token is not a header value: {source}")))?;
    value.set_sensitive(true);
    Ok(request.header(TOKEN_HEADER, value))
}

fn is_unauthorized(error: &Error) -> bool {
    matches!(error, Error::Api { status: 401, .. })
}

fn is_unauthorized_handshake(error: &Error) -> bool {
    matches!(
        error,
        Error::WebSocket(tungstenite::Error::Http(response)) if response.status() == 401
    )
}
